core/common/src/main/java/alluxio/resource/ResourcePool.java

```diff
@@ -69,9 +70,11 @@ protected ResourcePool(int maxCapacity, ConcurrentLinkedQueue<T> resources) {
    * the pool.
    *
    * @return a resource taken from the pool
+   * @throws IOException if an interruption occurs while acquiring resources
    */
-  public T acquire() {
-    return acquire(null, null);
+  @Override
+  public T acquire() throws IOException {
+    return acquire(WAIT_INDEFINITELY, null);
```

Comment:
It acts just like a flag, sorry for not quite understanding the 'complexity' means. Negative, then waits indefinitely, positive, then waits specifically, it can cover full domain of long and guarantee the code act as expectation.